File: sdk/gsusb-cpp/examples/03_simple_fd.cpp
#include "demo_common.hpp"

#include <iostream>

int main(int argc, char** argv) {
    try {
        const auto options = demo::parse_common_options(argc, argv, "Single-channel CAN FD example");
        auto tx_bus = std::make_unique<gsusb::GsUsbBus>(demo::make_bus_options(options, options.tx_channel, true));
        auto rx_bus = std::make_unique<gsusb::GsUsbBus>(demo::make_bus_options(options, options.rx_channel, true));

        gsusb::CanMessage message;
        message.arbitration_id = 0x1FFFFFF0u;
        message.is_extended_id = true;
        message.is_fd = true;
        message.bitrate_switch = true;
        message.data = demo::make_payload({0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88}, 64, 0x00);

        demo::print_message("Sending", message);
        tx_bus->send(message);
        auto received = rx_bus->recv(std::chrono::seconds(2));
        if (!received.has_value()) {
            std::cerr << "No FD frame received\n";
            return 1;
        }
        demo::print_message("Received", received.value());
        tx_bus->shutdown();
        rx_bus->shutdown();
        return 0;
    } catch (const std::exception& ex) {
        std::cerr << ex.what() << "\n";
        return 1;
    }
}

File: sdk/gsusb-csharp/examples/GsUsb.Examples/Examples/03_simple_fd.cs
using GsUsb;

namespace GsUsb.Examples;

internal static class Example03SimpleFd
{
    public static int Run(string[] args)
    {
        var options = CommonOptions.Parse(args, "Single-channel CAN FD example");
        using var txBus = new GsUsbBus(DemoCommon.MakeBusOptions(options, options.TxChannel, fd: true));
        using var rxBus = new GsUsbBus(DemoCommon.MakeBusOptions(options, options.RxChannel, fd: true));

        var message = new CanMessage
        {
            ArbitrationId = 0x1FFFFFF0u,
            IsExtendedId = true,
            IsFd = true,
            BitrateSwitch = true,
            Data = DemoCommon.MakePayload([0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88], 64, 0x00),
        };

        DemoCommon.PrintMessage("Sending", message);
        txBus.Send(message);
        var received = rxBus.Receive(TimeSpan.FromSeconds(2));
        if (received is null)
        {
            Console.Error.WriteLine("No FD frame received");
            return 1;
        }

        DemoCommon.PrintMessage("Received", received);
        return 0;
    }
}